T2D-297 — Retrieve reference sequence from ENA

tskir opened the pull request:
### [Diff link](https://github.com/EBIvariation/metadata-ws/pull/50/files)

* Implemented `ReferenceSequenceExtractor` and `ReferenceSequenceConverter classes`.
* Implemented `SraAssemblyXmlParser` class to parse the `ASSEMBLY` documents.
* Modified the application flow to use the new classes.

Points worth noting (& discussing):
* All created reference sequences have type == `ASSEMBLY` because this is the only type which ENA API ever returns
* All created reference sequences have patch == `NOT_SPECIFIED` because ENA data model never specifies the patch directly, only accession+version.

iyangar commented:
Did you fork the code from https://github.com/EBIvariation/metadata-ws/ to your local branch under `https://github.com/tskir`. It appears like you are directly working on upstream

iyangar commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/extractor/ReferenceSequenceExtractor.java:
```diff
@@ -0,0 +1,89 @@
+            String analysisXml = sraRetrieverByAccession.getXml(analysisAccession);
+            AnalysisType analysisType = sraAnalysisXmlParser.parseXml(analysisXml, analysisAccession);
+            String assemblyAccession = getAssemblyAccessionOfAnalysis(analysisType);
+            if (assemblyAccession == null) {return null;}  // not all analysis entities will have a reference sequence
```
formatting to next line

tskir replied:
Fixed: https://github.com/EBIvariation/metadata-ws/pull/50/commits/635dfb1c728c9281ee0de959ba86063fee170df0

iyangar commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/extractor/ReferenceSequenceExtractor.java:
```diff
@@ -0,0 +1,89 @@
+        String assemblyAccession = null;
+        ReferenceSequenceType referenceSequence = null;
+        AnalysisType.ANALYSISTYPE analysisType = analysis.getANALYSISTYPE();
+        if (analysisType.isSetREFERENCEALIGNMENT()) {referenceSequence = analysisType.getREFERENCEALIGNMENT();}
```
same as above

tskir replied:
Fixed

iyangar commented:
Please add test cases

selvaebi commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/extractor/ReferenceSequenceExtractor.java:
```diff
@@ -0,0 +1,89 @@
+            AnalysisType analysisType = sraAnalysisXmlParser.parseXml(analysisXml, analysisAccession);
+            String assemblyAccession = getAssemblyAccessionOfAnalysis(analysisType);
+            if (assemblyAccession == null) {return null;}  // not all analysis entities will have a reference sequence
+            String assemblyXml = sraRetrieverByAccession.getXml(assemblyAccession);
```
it only works through API , for database connections , in case of ega we need to fgure out what is the query to fetch , and also the accessions can be null in EGA case.

tskir replied:
Discussing with @cyenyxe my understanding was that the scope of this ticket is concerned only with retrieving the reference sequence from ENA by API, DB queries will have to be implemented separately

tskir replied:
Discussed this additionally with @selvaebi and @cyenyxe; decided that it doesn't make sense to implement reference sequence retrieval from the database, it will have to be done through API anyway.

tskir commented:
> Did you fork the code from https://github.com/EBIvariation/metadata-ws/ to your local branch under https://github.com/tskir. It appears like you are directly working on upstream https://github.com/EBIvariation/metadata-ws/tree/t2d-297-retrieve-reference-sequence

Yeah, it's just that at the time of starting work on this ticket I didn't know that forking is customary. This PR will have to be processed as is because GitHub does not allow to change the base branch, but all subsequent PRs from me will come from the fork.

tskir commented:
Earlier me and @selvaebi decided on the order of merge for several open PRs in metadata-load. It's going to be:
1. https://github.com/EBIvariation/metadata-ws/pull/48, T2D-288 — store study in metadata
2. https://github.com/EBIvariation/metadata-ws/pull/47, T2D-287 — store sample in metadata
3. https://github.com/EBIvariation/metadata-ws/pull/50, T2D-297 — Retrieve reference sequence from ENA (this PR)
4. https://github.com/EBIvariation/metadata-ws/pull/49, T2D-291 — Store sample in metadata

In line with this, I rebased this PR on top of T2D-287. To view only changes made by me and not by @selvaebi, this link will come in handy for reviewers:
https://github.com/selvaebi/metadata-ws/compare/T2D-287StoreAnalysis...ebivariation:t2d-297-retrieve-reference-sequence

selvaebi commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/extractor/ReferenceSequenceExtractor.java:
```diff
@@ -0,0 +1,93 @@
+        this.referenceSequenceConverter = referenceSequenceConverter;
+    }
+
+    public ReferenceSequence extractReferenceSequence(String analysisAccession) {
```
can you please move this code to SraObjectsImporterThroughApi

tskir replied:
Done in https://github.com/EBIvariation/metadata-ws/pull/50/commits/d504b323f7313c1800a3ba0d5872018c12644af0 and https://github.com/EBIvariation/metadata-ws/pull/50/commits/c8d1eb058b4838fbe58325440d5224047036a339

tskir commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Analysis.java:
```diff
@@ -105,7 +98,7 @@
             notes = "URL(s) to the reference sequence(s): either a single URL to an assembly/transcriptome, or a comma-separated list of gene sequences")
     @JsonProperty
-    @ManyToMany
+    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
```
As discussed with @selvaebi, this cascading needs to be ditched in favor of storing the reference sequence manually, in order to avoid duplicate issues

tskir replied:
Fixed in https://github.com/EBIvariation/metadata-ws/pull/50/commits/767b4d1a20eda626c547e83b812b3ea353078d00

tskir commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/ReferenceSequence.java:
```diff
@@ -87,11 +91,28 @@
     @ManyToOne(optional = false)
     private Taxonomy taxonomy;
```
As discussed with @selvaebi, taxonomy needs to be retrieved inside ReferenceSequence retrieval as well.

tskir replied:
Fixed in https://github.com/EBIvariation/metadata-ws/pull/50/commits/767b4d1a20eda626c547e83b812b3ea353078d00

tskir commented:
@iyangar 

> Please add test cases

Fixed in https://github.com/EBIvariation/metadata-ws/pull/50/commits/43b33e83988d8636aee8559f3cd29714f6666181.

tskir commented:
Rebased the code on top of the changes recently made by @selvaebi in #48 and #47.

Everything that was mentioned in the reviews is now fixed; **this PR is ready for the next iteration of review.** Merge order & handy diff link for reviewers are in this comment: https://github.com/EBIvariation/metadata-ws/pull/50#issuecomment-487908220

selvaebi commented:
we need to set the study taxonomy to be fetched from the reference sequence as well.

tskir commented:
> we need to set the study taxonomy to be fetched from the reference sequence as well.

Decided to make this a separate ticket during the sprint meeting because this is not related to the metadata import, but rather to the `metadata-ws` itself

tskir commented:
It was a bit painful (lots of changes happening!), but I finally rebased this branch on top of #47. It's now ready for the next iteration of review. Diff link: https://github.com/selvaebi/metadata-ws/compare/T2D-287StoreAnalysis...ebivariation:t2d-297-retrieve-reference-sequence

selvaebi commented:
Hi Kirill, sorry for that but, can you please rebase once again.

selvaebi commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/converter/ReferenceSequenceConverter.java:
```diff
@@ -0,0 +1,67 @@
+        return convertSaveAndReturnNewReferenceSequence(assemblyType, taxonomy);
+    }
+
+    private Taxonomy findExistingorSaveandReturnTaxonomy(AssemblyType assemblyType) {
```
Sorry I was wrong before can you please separate converter and persistence ,  like i have done in importStudy's dependentObjects.

tskir replied:
Fixed in https://github.com/EBIvariation/metadata-ws/pull/50/commits/6ff97f90c3c951061edceeeccc4855a94e78f12a. Now the three processes are separated:

* ReferenceSequence conversion
* Taxonomy extraction, lookup & persistence
* ReferenceSequence lookup & persistence

selvaebi commented on metadata-load/src/test/resources/assembly/assemblyDocumentAPI:
```diff
@@ -0,0 +1,266 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Can you rename the xmls and text files in CamelCase and with extension eg: AssemblyDocumentAPI.xml

tskir replied:
Fixed in https://github.com/EBIvariation/metadata-ws/pull/50/commits/1847b04057735e88ea0a3fe099dc559c01405dd8.

tskir commented:
> Hi Kirill, sorry for that but, can you please rebase once again.

Rebased against the latest changes #47. Fixed all other comments. **This request is now ready for the 4th iteration of review.**

[*Handy diff link which includes only changes from this PR*](https://github.com/selvaebi/metadata-ws/compare/T2D-287StoreAnalysis...ebivariation:t2d-297-retrieve-reference-sequence)

selvaebi commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/repositories/ReferenceSequenceRepository.java:
```diff
@@ -36,4 +39,6 @@ default void customize(QuerydslBindings bindings, QReferenceSequence referenceSe
                 .first((path, value) -> path.any().equalsIgnoreCase(value.iterator().next()));
     }
 
+    ReferenceSequence findByAccessions(@Param("accessions") List<String> accessions);
```
does this query return only one reference-sequence?

tskir replied:
Discussed this during yesterday's standup; decided to leave as is for now. Natural key for the reference key is going to be the list of its accessions. Because of the way it's implemented, reference sequences which have exactly the same list of accessions will not be duplicated, hence this will always return only one reference sequence.

selvaebi commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/converter/ReferenceSequenceConverter.java:
```diff
@@ -0,0 +1,41 @@
+
+public class ReferenceSequenceConverter implements Converter<AssemblyType, ReferenceSequence> {
+
+    public ReferenceSequenceConverter() {
```
No need for this constructor

tskir replied:
Fixed in https://github.com/EBIvariation/metadata-ws/pull/50/commits/1237be8d5509f5705b777378071b4c960f623511

iyangar commented:
General: metadata-ws module has cucumber test cases. Should tests here be continued as is or converted to cucumber later

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/database/MetadataReferenceSequenceFinderOrPersister.java:
```diff
@@ -0,0 +1,47 @@
+        /* The below find query will make sure to return the same reference sequence when analyses sharing same
+        reference sequence are imported in different runs */
+        ReferenceSequence existingReferenceSequence = referenceSequenceRepository.findByAccessions(
+                referenceSequence.getAccessions());
```
As discussed during the standup yesterday, currently the ReferenceSequence is identified by a full list of its accessions. In the future a more fitting & complicated solution will likely need to be implemented, in a separate ticket. For initial import this solution will work because all identical reference sequences will use the same list of accessions.

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/database/MetadataStudyFinderOrPersister.java:
```diff
@@ -0,0 +1,50 @@
+        if (sharedStudy != null) {
+            return sharedStudy;
+        }
+        return persistStudy(study);
```
Sorry, I'm not sure if I understood your concern correctly. Currently there are two DB queries: first attempts to find an existing Study, second persists it if necessary. Without the first query we can't be sure that the same Study hasn't already been imported which will lead to duplicates. Besides, the code for `MetadataStudyFinderOrPersister.java` is @selvaebi's, not mine :)

tskir commented:
Rebased on top of recent changes in #47. Addressed review comments. **Ready for the fifth iteration of review.** [*Diff link*](https://github.com/selvaebi/metadata-ws/compare/T2D-287StoreAnalysis...ebivariation:t2d-297-retrieve-reference-sequence)

selvaebi commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/ObjectsImporter.java:
```diff
@@ -142,6 +162,55 @@ public Analysis importAnalysis(String accession) {
+        // REFERENCE_ALIGNMENT, SEQUENCE_VARIATION, and PROCESSED_READS. All accessions are collected and returned.
+        if (analysisType1.isSetREFERENCEALIGNMENT()) {
+            referenceSequenceAccessions.add(
+                    // TODO: should we use getRefname() or getAccession()?
```
All EGA xmls have refName only

selvaebi replied:
so I guess we can just use the refName to create the referenceSequence in SraObjectImportThroughDatabase

tskir replied:
@selvaebi Thank you for the clarification. I changed the implementation so that `getAccession()` is used by default while `getRefname()` is used for import from the database.

@cyenyxe What should we do in the case of ENA? Consider the two following examples:
* Analysis [ERZ496533](https://www.ebi.ac.uk/ena/data/view/ERZ496533&display=xml), which is a `SEQUENCE_VARIATION`, has assembly [GCA_000002305.1](https://www.ebi.ac.uk/ena/data/view/GCA_000002305.1&display=xml) which only has an accession. This is a standard assembly which can be parsed & used as usual.
* Analysis [ERZ004000](https://www.ebi.ac.uk/ena/data/view/ERZ004000&display=xml), which is a `REFERENCE_ALIGNMENT`, has assembly [HAAA01000000](https://www.ebi.ac.uk/ena/data/view/HAAA01000000&display=xml), which only has a refName. Moreover, this is not a standard assembly; it returns an <ENTRY> entity, and I'm not sure what to do with them (and whether we _want_ to do anything with them)

cyenyxe replied:
We should only process `SEQUENCE_VARIATION` analyses for now, but your question about HAAA01000000 still applies.

That is a transcriptome, one of the 3 types of sequences we support, so it has to be processed. The set of sequences it contains is
> \<xref db="ENA-TSA" id="HAAA01000001-HAAA01003524"/\>

tskir replied:
@iyangar Transcriptome and gene sequences are still not supported in the latest version of the PR. See example of a transcriptome reference sequence here:
* https://www.ebi.ac.uk/ena/data/view/ERZ004000&display=xml
* https://www.ebi.ac.uk/ena/data/view/HAAA01000000&display=xml

For it, we have an empty `STANDARD` (no accession, only a refname), and the actual accessions are directly inside the `REFERENCE_ALIGNMENT`:

```
<SEQUENCE accession="HAAA01000001.1" label="PNEJI2.1.1"/>
<SEQUENCE accession="HAAA01000002.1" label="PNEJI2.2.1"/>
<SEQUENCE accession="HAAA01000003.1" label="PNEJI2.3.1"/>
...
```

iyangar replied:
correct, In today's standup we decided to discuss about this in sprint meeting

selvaebi commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/configuration/MetadataImporterMainApplicationConfiguration.java:
```diff
@@ -47,34 +51,59 @@
-                webResourceExtractorFromStudy(webResourceRepository), taxonomyExtractor(taxonomyRepository),
-                sraAnalysisXmlParser(), analysisConverter(), fileExtractorFromAnalysis(fileRepository),
-                analysisRepository,studyRepository);
+    public ObjectsImporter objectImporterThroughEnaDatabase(SraXmlRetrieverThroughDatabase sraXmlRetrieverThroughDatabase,
```
SraXmlRetrieverThroughDatabase can be moved to the next line.

tskir replied:
Fixed, thank you

selvaebi commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/repositories/ReferenceSequenceCustomRepository.java:
```diff
@@ -0,0 +1,44 @@
+    ReferenceSequence findByAccessions(@Param("accessions") List<String> accessions);
+
+    default ReferenceSequence findOrSave(ReferenceSequence referenceSequence) {
+        /* The below find query will make sure to return the same reference sequence when analyses sharing same
```
i guess we can remove the comment as metadata-ws doesn't need to know about import here.

tskir replied:
Makes sense, done

selvaebi commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/repositories/ReferenceSequenceCustomRepository.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.List;
+
+@NoRepositoryBean
+public interface ReferenceSequenceCustomRepository extends PagingAndSortingRepository<ReferenceSequence, Long>,
```
```suggestion
public interface ReferenceSequenceRepositoryCustom extends PagingAndSortingRepository<ReferenceSequence, Long>,
```

tskir replied:
Renamed, thanks

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/MetadataImporterMainApplication.java:
```diff
@@ -45,22 +44,23 @@
 
     public static void main(String[] args) throws Exception {
         SpringApplication.run(MetadataImporterMainApplication.class, args);
+
```
Done

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/ObjectsImporter.java:
```diff
@@ -142,6 +162,55 @@ public Analysis importAnalysis(String accession) {
+                    // TODO: should we use getRefname() or getAccession()?
+                    getAccessionFromReferenceSequenceType(analysisType1.getREFERENCEALIGNMENT()));
+        }
+        else if (analysisType1.isSetSEQUENCEVARIATION()) {
```
Noted & fixed, thank you

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/ObjectsImporter.java:
```diff
@@ -142,6 +162,55 @@ public Analysis importAnalysis(String accession) {
+        return referenceSequenceAccessions;
+    }
+
+    private String getAccessionFromReferenceSequenceType(ReferenceSequenceType referenceSequenceType){
```
Fixed

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/ObjectsImporter.java:
```diff
@@ -142,6 +162,55 @@ public Analysis importAnalysis(String accession) {
+            referenceSequenceAccessions.add(
+                    getAccessionFromReferenceSequenceType(analysisType1.getPROCESSEDREADS()));
+        }
+        referenceSequenceAccessions.remove(null);
```
Agreed; and I also noticed an implementation flaw in that is should have been not `if / else if / else if` block, but rather just a series of `if`s. I'm not sure whether an ENA analysis can include multiple types in practice, but the parser class implementation (e. g. method is named `isSetREFERENCEALIGNMENT`, not `isREFERENCEALIGNMENT`) suggests that theoretically it can happen.

Rewrote this block to be a series of `if`s and also to avoid removing `null` afterwards

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/configuration/MetadataImporterMainApplicationConfiguration.java:
```diff
@@ -47,34 +51,59 @@
-                webResourceExtractorFromStudy(webResourceRepository), taxonomyExtractor(taxonomyRepository),
-                sraAnalysisXmlParser(), analysisConverter(), fileExtractorFromAnalysis(fileRepository),
-                analysisRepository, studyRepository);
+                                                   PublicationRepository publicationRepository,
```
Hmm, it's really strange I didn't notice this. Now fixed!

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/configuration/MetadataImporterMainApplicationConfiguration.java:
```diff
@@ -47,34 +51,59 @@
+        );
     }
 
     private StudyConverter studyConverter() {
```
I noticed this problem as well and have already performed a conceptual grouping in #55. I suggest that in this PR we leave this as is to avoid unnecessary downstream rebasing.

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/converter/ReferenceSequenceConverter.java:
```diff
@@ -0,0 +1,38 @@
+/*
+ *
+ * Copyright 2018 EMBL - European Bioinformatics Institute
```
Fixed. Maybe we should do a global replace in all files to make it "2018-2019"?

cyenyxe replied:
Not really, given that we copyright them individually, each of them should start in the creation date.

tskir replied:
Then, presumably, we also need to update the years for modules which were created in 2018 and then edited in 2019, i. e. “Copyright 2018” → “Copyright 2018–2019”? I suspect we'll generally forget to maintain this :)

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/converter/ReferenceSequenceConverter.java:
```diff
@@ -0,0 +1,38 @@
+                assemblyType.getNAME(),
+                "NOT_SPECIFIED",  // ENA only specifies accession+version, not the patch
+                Arrays.asList(assemblyType.getAccession()),
+                ReferenceSequence.Type.ASSEMBLY  // ENA data model only has ASSEMBLY type
```
I've looked through the entire `uk.ac.ebi.ena.sra.xml` package but could't find anything resembling these types. I also compiled a [spreadsheet](https://docs.google.com/spreadsheets/d/18IT7defkFXC0zkwhO0kQyGWOi1UbMmN9SjJq6B3yHfc/edit?usp=sharing) of all possible analysis types with accession examples, and in my experience they either contain a `ReferenceSequenceType`, which in turn links to an `ASSEMBLYType` (and _only_ to it), or they contain nothing at all. If you could provide me with a pointer, e. g. an accession example where you think the reference sequence is not an assembly, but instead a gene sequence or a transcriptome, I would be thankful.

This may be related to the refName/accession thing (see discussion [here](https://github.com/EBIvariation/metadata-ws/pull/50#discussion_r287836865)), because one of the analysis accessions returns something weird which is not an Assembly (https://www.ebi.ac.uk/ena/data/view/HAAA01000000&display=xml). May this be what we're looking for?

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/extractor/TaxonomyExtractorFromReferenceSequence.java:
```diff
@@ -0,0 +1,46 @@
+import uk.ac.ebi.ampt2d.metadata.persistence.repositories.TaxonomyRepository;
+import uk.ac.ebi.ena.sra.xml.AssemblyType;
+
+public class TaxonomyExtractorFromReferenceSequence {
```
I can't _yet_ rename it to `TaxonomyExtractor` because this name is taken by a dummy taxonomy generator, which will only be removed when we merge #54.

As for the class structure:
* The taxonomy extraction & persistence was originally implemented as a separate class;
* Then, following review comments and changing upstream class structure, moved inside the ReferenceSequence persistence logic;
* Then moved to a separate class _again_ as per @selvaebi's suggestion.

Currently it is modelled after `PublicationExtractorFromStudy` and similar other classes. Could you please elaborate on how I should change it and where to move its code?

cyenyxe replied:
I guess my concern is due to the fact that this class doesn't really deal with a reference sequence at any point, just with taxonomies, and that is why I see it been closer to `StudyRepository`. If that was the case, we would need to replicate the `StudyRepositoryCustom(Impl)` class hierarchy, the `findOrCreate` method in the implementation instead of the top interface declaration.

tskir replied:
@cyenyxe I see your point now. I suppose the best way would be to implement a `TaxonomyRepositoryCustom` with  a `findOrCreate` method and to simply call that at the appropriate point during the reference sequence import

tskir commented on metadata-load/src/test/java/uk/ac/ebi/ampt2d/metadata/importer/converter/ReferenceSequenceConverterTest.java:
```diff
@@ -0,0 +1,82 @@
+        String assemblyAccession = "GCA_000002305.1";
+        AssemblyType assemblyType = getAssemblyType(ASSEMBLY_DOCUMENT_API_XML, assemblyAccession);
+        ReferenceSequence referenceSequence = referenceSequenceConverter.convert(assemblyType);
+        Assert.assertNotNull(referenceSequence);
```
Fixed

tskir commented:
Addressed all review comments. **Ready for the sixth iteration of review.** [Diff link](https://github.com/EBIvariation/metadata-ws/pull/50/files)

cyenyxe commented:
Have the changes been pushed? I still see the old version of the code.

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/ObjectsImporter.java:
```diff
@@ -142,6 +162,55 @@ public Analysis importAnalysis(String accession) {
+    }
+
+    private String getAccessionFromReferenceSequenceType(ReferenceSequenceType referenceSequenceType){
+        if (referenceSequenceType == null) {
```
Thanks, fixed

tskir commented:
> Have the changes been pushed? I still see the old version of the code.

Yep, I indeed forgot to push it >< Now fixed as well as the latest suggestion from @iyangar. Ready for review

cyenyxe commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/MetadataImporterMainApplication.java:
```diff
@@ -52,7 +52,6 @@ public MetadataImporterMainApplication(ObjectsImporter objectsImporter) {
     public static void main(String[] args) throws Exception {
         SpringApplication.run(MetadataImporterMainApplication.class, args);
     }
```
```suggestion
    }

```

tskir replied:
Thanks, fixed

cyenyxe commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/ObjectsImporter.java:
```diff
@@ -142,9 +162,64 @@ public Analysis importAnalysis(String accession) {
+        Set<String> referenceSequenceAccessions = new HashSet<>();
+        AnalysisType.ANALYSISTYPE analysisType1 = analysisType.getANALYSISTYPE();
+        // ENA records can contain a reference sequence in either of the three analysis categories:
+        // REFERENCE_ALIGNMENT, SEQUENCE_VARIATION, and PROCESSED_READS. All accessions are collected and returned.
```
The EVA and AMP only deal with SEQUENCE_VARIATION analyses, could @jdylan please confirm if EGA needs others?

tskir replied:
By the way, should metadata API support ENA as well eventually, or it's not planned, ever?

cyenyxe replied:
The API doesn't need to support ENA.

iyangar replied:
Will remove `REFERENCE_ALIGNMENT` and `PROCESSED_READS`

cyenyxe replied:
Has @jdylan confirmed that the EGA is not using them?

iyangar replied:
I interpreted https://github.com/EBIvariation/metadata-ws/pull/50#discussion_r288970412 so. Will wait for formal confirmation from @jdylan

selvaebi replied:
I could see from the xmls EGA only uses SEQUENCE_VARIATION. but anyways Dylan can confirm.

iyangar replied:
Need to handle all 3. https://www.ebi.ac.uk/panda/jira/browse/T2D-297?focusedCommentId=281856&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-281856

tskir replied:
In the latest version of the PR it's implemented slightly different from my original implementation. Originally it combined accessions from all types of analyses if it could find them; now it checks sequentially in a `else-if` chain. If we are guaranteed for sure that two different tags cannot be present in the same analysis (e. g. a `SEQUENCE_VARIATION` and a `REFERENCE_ALIGNMENT`), then it's fine. @cyenyxe @jdylan can you confirm that each analysis contains at most one analysis type?

iyangar replied:
Yes, in today's i confirmed with @jdylan. Analysis can be of only one type SEQUENCE_VARIATION or REFERENCE_ALIGNMENT or PROCESSED_READS

tskir replied:
In the current implementation all three analysis types are supported in a reasonable manner. Marking as resolved.

cyenyxe commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/database/SraObjectsImporterThroughDatabase.java:
```diff
@@ -118,6 +119,11 @@ protected Study extractAnalysisFromStudy(StudyType studyType, Study study) {
     }
 
+    @Override
+    protected String getAccessionFromStandard(ReferenceAssemblyType.STANDARD standard) {
```
It is counter-intuitive that a method expected to return the accession in fact doesn't. What is the reason to return always the refname when importing from the database?

selvaebi replied:
it should be for EGA case, maybe if the object has accession call super. import else, need to create RefSeq from refname.

tskir replied:
> It is counter-intuitive that a method expected to return the accession in fact doesn't

It is, but the reason we have to do this, again if I understand correctly, because sometimes refName is used as a _de facto_ accession in EGA, and we have to use it to identify the reference sequence. Maybe @selvaebi and @jdylan could either confirm or correct me.

cyenyxe replied:
Even if EGA uses the reference name right now, that can be easily mapped to a proper accession, for instance:

* GRCh37 would have the accession GCA_000001405.1
* GRCh38 would have the accession GCA_000001405.15
* GRCh38.p13 (latest) would have the accession GCA_000001405.28

We should always favor accessions because they are guaranteed to be unique.

iyangar replied:
Are we confined to only above three vi.z., GRCh37, GRCh38 and GRCh38.p13. 
Should we consider other reference name like `GRCh37.p13` with accession `GCA_000001405.14`

cyenyxe replied:
Yes, we need to consider from GCA_000001405.1 to GCA_000001405.28.

iyangar replied:
Mapping case-sensitive reference name to accession

- GRCh37 through GRCh37.p13 = GCA_000001405.1 through GCA_000001405.14

- GRCh38 through GRCh38.p13 = GCA_000001405.15 through GCA_000001405.28

Ref: https://www.ncbi.nlm.nih.gov/assembly/?term=GRCh37 and https://www.ncbi.nlm.nih.gov/assembly?term=GRCh38&cmd=DetailsSearch

tskir replied:
This has been fixed & implemented by @iyangar. Now certain human build refnames are converted to accessions

iyangar commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/ObjectsImporter.java:
```diff
@@ -142,9 +162,64 @@ public Analysis importAnalysis(String accession) {
     }
 
+    protected String getAccessionFromStandard(ReferenceAssemblyType.STANDARD standard) {
+        return standard.getAccession();
```
@cyenyxe just wanna confirm this is fine? 
[ObjectImporter::getAccessionFromStandard() and SraObjectsImporterThroughDatabase::getAccessionFromStandard()   https://github.com/EBIvariation/metadata-ws/pull/50#discussion_r288505075 ]

cyenyxe replied:
Yes, it must always return the accession.

tskir replied:
In the current implementation accession is always returned. Marking as resolved.

cyenyxe commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/database/SraObjectsImporterThroughDatabase.java:
```diff
@@ -121,7 +125,131 @@ protected Study extractAnalysisFromStudy(StudyType studyType, Study study) {
+        String accession = null;
+
+        try {
+            switch (refName) {
```
Instead of making this effectively a linear search through the use of a switch, you can create a static (initialized just once) Map and store the mappings there. That will make every search have O(1) complexity.

I would also suggest making the search case insensitive, we have seen `grch37` or similar used many times.

tskir replied:
Both points fixed by @iyangar

cyenyxe commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/database/SraObjectsImporterThroughDatabase.java:
```diff
@@ -126,130 +157,7 @@ protected Study extractAnalysisFromStudy(StudyType studyType, Study study) {
-                    break;
-
-                default:
-                    throw new Exception("Encountered exception for unknown reference sequence name " + refName);
```
This exception still applies, using a Map doesn't prevent the user from requesting an unknown reference. But it should be an `IllegalArgumentException` instead of a completely generic `Exception`.

tskir replied:
Fixed by @iyangar; in the current implementation `IllegalArgumentException` is thrown

selvaebi commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/database/SraObjectsImporterThroughDatabase.java:
```diff
@@ -126,130 +157,7 @@ protected Study extractAnalysisFromStudy(StudyType studyType, Study study) {
-        }
-
-        return accession;
+        return REFERENCE_MAP.get(refName.toLowerCase());
```
please add a check for refName !=null

tskir replied:
Fixed by @iyangar, check is now present

selvaebi commented:
please add test cases as well.

selvaebi commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/ObjectsImporter.java:
```diff
@@ -159,6 +199,41 @@ public Analysis importAnalysis(String accession) {
+                        String refName = standard.getRefname();
+                        String accession = null;
+                        if (refName != null) {
+                            accession = REFERENCE_MAP.get(refName.toLowerCase());
```
this manual creation is just for database case / ega, so it should be moved to SraDatabaseImporter class

tskir replied:
Fixed by @iyangar, now moved to `SraObjectsImporterThroughDatabase`

selvaebi commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/ObjectsImporter.java:
```diff
@@ -158,20 +238,38 @@ public Sample importSample(String accession) {
+        if (referenceSequenceType != null) {
+            ReferenceAssemblyType referenceAssemblyType = referenceSequenceType.getASSEMBLY();
+            if (referenceAssemblyType != null) {
+                ReferenceAssemblyType.STANDARD standard = referenceAssemblyType.getSTANDARD();
```
This all logic of manual creation of ReferenceSequnce should go to SraImporterDatabase

tskir replied:
This comment was probably made before @iyangar made his latest changes (I'm not sure). Anyway in the latest implementation `getAccessionFromReferenceSequenceType` method should remain in the `ObjectsImporter` class because it is used in both API and database cases. It is used to extract accession from ENA's `ReferenceSequenceType`.

selvaebi commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/ObjectsImporter.java:
```diff
@@ -118,12 +140,15 @@ public Analysis importAnalysis(String accession) {
-            for (String referenceSequenceAccession : getReferenceSequenceAccessions(analysisType)) {
-                //TODO ReferenceSequence Import
-                //referenceSequences.add(importReferenceSequence(referenceSequenceAccession));
+            ReferenceSequence referenceSequence = buildReferenceSequence(analysisType);
```
ImportReferenceSequence is not called at all  ?

iyangar replied:
In ImportReferenceSequence() => IsraXmlRetrieverByAccession.getXml is selecting analysis record with reference sequence accession and fails. If i am correct on slack we discussed this and I interpreted DB call should not be for ref seq. Please let me know

selvaebi replied:
yes, you need to override importReferenceSequence in SraImporterDatabase to manually create refSequence instead calling super.

iyangar replied:
sure

iyangar replied:
Is there a way to get reference sequence name from its accession? (From analysisType ref seq accession is retrieved and getting ref name is inside this is straight forward)

tskir replied:
@iyangar it should be possible, although accessions are not always present and they are not always RefSeq accessions. Missing refname is one of the edge cases which I outlined in the [spreadsheet](https://docs.google.com/spreadsheets/d/18IT7defkFXC0zkwhO0kQyGWOi1UbMmN9SjJq6B3yHfc/edit?usp=sharing), let's see what decision is made concerning it.

iyangar replied:
This is implemented in SraObjectsImporterThroughDatabase. This comment was made before i implemented by saving both accession and ref name.

selvaebi commented on metadata-load/src/test/java/uk/ac/ebi/ampt2d/metadata/importer/api/SraObjectsImporterThroughAPITest.java:
```diff
@@ -94,4 +99,16 @@ public void importAnalysisObject() throws Exception {
\ No newline at end of file
+
+    @Test
+    public void importReferenceSequenceObject() throws Exception {
```
also, add asserts in importStudy to check whether reference sequences are imported?

selvaebi replied:
this is not done yet

tskir replied:
Actually, currently no reference sequences are imported in `importStudy` , because two studies don't have any analyses (SRP000392, SRP000118), and the remaining study ERP000054 has only one analysis ERZ000001 which has a custom (currently non-importable) reference sequence:
```
               <ASSEMBLY>
                    <CUSTOM>
                         <URL_LINK>
                              <LABEL>Mus_musculus.NCBIM37.52</LABEL>
                              <URL>ftp://ftp.ensembl.org/pub/release-52/fasta/mus_musculus/dna/Mus_musculus.NCBIM37.52.dna.toplevel.fa.gz</URL>
                         </URL_LINK>
                    </CUSTOM>
               </ASSEMBLY>
```

To add this test, I replaced `ERP000054` with `ERP006576`, which has two analyses and one reference sequence.

tskir commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/ObjectsImporter.java:
```diff
@@ -22,7 +22,6 @@
 import uk.ac.ebi.ampt2d.metadata.importer.extractor.FileExtractorFromAnalysis;
```
This is a valid concern. I suggest we move it to #55, because I have already performed a significant reordering/grouping of methods and class attributes in that PR. If I change anything here, an unnecessary chain of downstream rebases would be required.

cyenyxe replied:
We don't order methods as in public, protected then private. We follow the recommendations from the Clean Code book, where a private method follows the first public method that uses it, in order to keep a particular flow as close as possible.

tskir replied:
Fixed in #55, see [discussion](https://github.com/EBIvariation/metadata-ws/pull/55#pullrequestreview-247590754)

cyenyxe commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/database/SraObjectsImporterThroughDatabase.java:
```diff
@@ -76,7 +137,20 @@ public Analysis importAnalysis(String accession) {
     @Override
     public ReferenceSequence importReferenceSequence(String accession) {
-        return super.importReferenceSequence(accession);
+        String[] accnRefNameSplit = accession.split("#", 2);
```
The split by '#' is not very intuitive (although documented in another method). Could you please add some method documentation to show what the input of this method looks like, and where does it come from?

tskir replied:
Added the documentation in https://github.com/EBIvariation/metadata-ws/pull/50/commits/f992ccc9192cdef417bcfafd9a2ea1b554e32493. This is @iyangar's implementation for refname/accession handling in EGA import, and I think it's pretty reasonable

cyenyxe commented on metadata-load/src/main/java/uk/ac/ebi/ampt2d/metadata/importer/database/SraObjectsImporterThroughDatabase.java:
```diff
@@ -76,7 +137,20 @@ public Analysis importAnalysis(String accession) {
+        if (refNameSplit.length == 2) {
+            patch = refNameSplit[1];
+        }
+        ReferenceSequence referenceSequence = new ReferenceSequence(refName, patch, accessionList,  ReferenceSequence.Type.ASSEMBLY);
```
```suggestion
        ReferenceSequence referenceSequence = new ReferenceSequence(refName, patch, accessionList, ReferenceSequence.Type.ASSEMBLY);
```

tskir replied:
Fixed in https://github.com/EBIvariation/metadata-ws/pull/50/commits/f992ccc9192cdef417bcfafd9a2ea1b554e32493.

tskir commented:
> According to GitHub it looks like it wasn't fixed, could you please edit it manually?

Whoops, turned out I fixed this, but in the wrong PR (one of the downstream ones). Now fixed here as well, commit https://github.com/EBIvariation/metadata-ws/pull/50/commits/f992ccc9192cdef417bcfafd9a2ea1b554e32493.
